### PR TITLE
Moderator api

### PR DESCRIPTION
**Pull request by Martinfree**
2 tasks:

1. https://trello.com/c/z0lxajL3/108-api-%D0%BC%D0%BE%D0%B4%D0%B5%D1%80%D0%B0%D1%82%D0%BE%D1%80%D0%B0

2. https://trello.com/c/fdgp5mJ0/109-api-%D0%B0%D0%B4%D0%BC%D1%96%D0%BD%D0%B0

### CODE REVIEW

File: api/utils/decorators.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 def permission(permission):
     def perm(func):
         def p(request,args,**kwargs):
-            permis = P.pop(permission)
+            permis = P.get(permission)
```
Comment (klekhaav) [inline api/utils/decorators.py:15]: cool

File: api/utils/models.py [hunk 2]
```diff
@@ -1,3 +1,39 @@
...
+from django.core.mail import send_mail
 from django.db import models
+from django.utils import timezone
+from django.core.signing import TimestampSigner, b64_encode,b64_decode, BadSignature, SignatureExpired,force_bytes
+from django.core.mail import EmailMessage
+from django.conf import settings 
+
+from ext_news.models import News
+
+class Mail(models.Model):
```
Comment (klekhaav) [inline api/utils/models.py]: do we need extra class for simple mail sender?

File: api/settings/base.py [hunk 3]
```diff
@@ -37,8 +37,11 @@
     'django_nose',
 
     'djangotoolbox',
+    
     'authentication',
     'ext_news',
+    'utils',
```
Comment (klekhaav) [inline api/settings/base.py]: do we need utils as app? we can use it as usual utils

File: api/utils/serializers.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+from django.contrib.auth.password_validation import validate_password
```
Comment (klekhaav) [inline api/utils/serializers.py]: question about app

File: api/utils/views.py [hunk 5]
```diff
@@ -1,3 +1,90 @@
...
+    SendMailSerializer,
+    MassMailSerializer,
+   )
+from utils.decorators import permission, permissions, object_permission
+
+User = get_user_model()
+
+
+
+class SendMailAPIView(APIView):
```
Comment (klekhaav) [inline api/utils/views.py:28]: we can use function based call for send mail

Comment (klekhaav) [reply api/utils/views.py:28]: but i like how you made that

File: api/utils/decorators.py (hunk 1 above)
Comment (Martinfree) [reply api/utils/decorators.py:15]: Redo or cool?

**Review by klekhaav — COMMENTED** [review]
i like it

File: api/utils/serializers.py (hunk 4 above)
Comment (Martinfree) [reply api/utils/serializers.py]: that's my little copy/paste (resolve)

File: api/utils/models.py (hunk 2 above)
Comment (Martinfree) [reply api/utils/models.py]: already done

File: api/settings/base.py (hunk 3 above)
Comment (Martinfree) [reply api/settings/base.py]: deleted for now

File: api/authentication/views.py [hunk 6]
```diff
@@ -81,7 +82,7 @@ def put(self, request, *args, **kwargs):
         queryset = User.objects.filter(id=number.get('id'))
         serializer = UpdateUserSerializer(queryset[0],  data=request.data)
         if serializer.is_valid():
-            serializer.save()
+            serilizer.save()
```
Comment (pavlo-mk) [inline api/authentication/views.py]: I think you should save `serializer` object.

File: api/authentication/views.py [hunk 7]
```diff
@@ -361,3 +345,64 @@ class UserInactiveAPIView(APIView):
...
     def post(self, request, **kwargs):
         return Response({"Account inactive!"}, status=status.HTTP_400_BAD_REQUEST)
+
+class SetModeratorAPIView(APIView):
+    permission_classes = [AllowAny, ]
+    queryset = User.objects.none()
+    
+    @permission('IsAdminUser') 
+    def post(self, request, *args, **kwargs):
+        self.serializer_class = SetModeratorSerializer
```
Comment (pavlo-mk) [inline api/authentication/views.py]: if you use `serializer_class` please add it in mail (parent) class above and later in methods you can call only `self.serializer_class`.  You have in this place like rewrite main option but you didn't added it in parent class.

File: api/authentication/views.py [hunk 8]
```diff
@@ -361,3 +345,64 @@ class UserInactiveAPIView(APIView):
...
+                    status=status.HTTP_200_OK)
+
+class AdminUserAPIView(APIView):
+    permission_classes = [AllowAny, ]
+    queryset = User.objects.none()
+
+    @permission("IsAdminUser")
+    def post(self, request,*args,**kwargs):
+        self.serializer_class = CreateUserSerializer    
+        serializer = self.serializer_class(data=request.data)
```
Comment (pavlo-mk) [inline api/authentication/views.py]: the same as above. move `self.serializer_class` to class field.

File: api/authentication/views.py [hunk 9]
```diff
@@ -361,3 +345,64 @@ class UserInactiveAPIView(APIView):
...
+        user.save()
+        return Response({'Status':'OK'},status=status.HTTP_200_OK)
+
+class BanUserAPIView(APIView):
+    permission_classes = [AllowAny, ]
+    queryset = User.objects.none()
+
+    @permission("IsModeratorUser")
+    def post(self,request,*args,**kwargs):
+        self.serializer_class = SetModeratorSerializer
```
Comment (pavlo-mk) [inline api/authentication/views.py]: the same

**Review by pavlo-mk — CHANGES_REQUESTED** [review]
LGTM😊👍. Couple small changes requested. And can merge if all good work)!

**Pull request merged**